feat: add `no-useless-assignment` rule

ota-meshi opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[X] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This PR adds a new `no-useless-assignment` rule which warns if a variable is not used after being assigned a value.

close #17559

#### Is there anything you'd like reviewers to focus on?

~~This PR is still a draft as many test cases are missing.~~

<!-- markdownlint-disable-file MD004 -->

ota-meshi commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,284 @@
+            onUnreachableCodePathSegmentEnd() {
+                scopeStack.currentSegments.shift();
+            },
+            Identifier(node) {
+                const segmentId = scopeStack.currentSegments[0].id;
+
+                scopeStack.segmentIdentifiers[segmentId].add(node);
+            },
```
Hmm... I think this is probably not the best way to do it.

mdjermanovic commented on tests/lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,564 @@
+        foo = 'unused like but exported';`,
+
+        // Basic tests
+        `let v = 'used';
+        console.log(v);
+        v = 'used-2'
+        console.log(v);`,
```
It also doesn't report when the value is overwritten before being read after an assignment:

```js
/* eslint no-useless-assignment: 2 */

let v;
v = 'unused'; // false negative?
v = 'used';
console.log(v);

```

ota-meshi replied:
@mdjermanovic 
If you already have an idea to implement it please let me know 🙏

mdjermanovic replied:
> If you already have an idea to implement it please let me know

I don't have any specific idea at the moment, but I imagine it would be somewhat similar to rules that use code path analysis to check if something was executed in any/all preceding code paths, for example `no-this-before-super`.

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,419 @@
+/** @type {import('../shared/types').Rule} */
+module.exports = {
+    meta: {
+        type: "suggestion",
```
```suggestion
        type: "problem",
```

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,419 @@
+        type: "suggestion",
+
+        docs: {
+            description: "Disallow variables that are not used after assignment",
```
```suggestion
            description: "Disallow variable assignments when the value is not used",
```

ota-meshi commented:
Hmm. I don't understand why it crashes in CI. I can't reproduce it well locally. If you know the cause, could you please advise?

mdjermanovic commented:
> Hmm. I don't understand why it crashes in CI. I can't reproduce it well locally. If you know the cause, could you please advise?

The problem is fixed in dependencies now (https://github.com/eslint/eslint/pull/17664#issuecomment-1771457180, fixed by https://github.com/humanwhocodes/object-schema/commit/6ca80b001a4ffb678b9b5544fc53322117374376).

mdjermanovic commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,458 @@
+
+                if (!readReferences.length) {
+
+                    // It is unused variable
+                    return;
```
This is to avoid reporting on variables that are reported by `no-unused-vars`?

ota-meshi replied:
I thought it shouldn't be reported in the new rule because it's not an unnecessary assignment, it's clearly an unnecessary variable. And it is reported by `no-unused-vars`.

nzakas replied:
Can you add those details both as a comment here and in the documentation? I can see this being a bit confusing, but I agree with your rationale.

mdjermanovic commented:
I tried enabling this rule in the eslint codebase (this branch) and it found 46 errors:

<details>

```
C:\projects\tmp\ota-meshi-eslint\Makefile.js
  375:9  error  This assigned value is not used in subsequent statements  no-useless-assignment
  388:9  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\linter\code-path-analysis\code-path.js
  183:13  error  This assigned value is not used in subsequent statements  no-useless-assignment
  184:13  error  This assigned value is not used in subsequent statements  no-useless-assignment
  185:13  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\linter\config-comment-parser.js
   75:13  error  This assigned value is not used in subsequent statements  no-useless-assignment
  102:9   error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\linter\linter.js
  2062:13  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\linter\timing.js
  140:17  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\array-bracket-newline.js
  74:17  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\indent-legacy.js
  712:17  error  This assigned value is not used in subsequent statements  no-useless-assignment
  833:17  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\indent.js
  1407:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\key-spacing.js
  489:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\lines-around-comment.js
  156:17  error  This assigned value is not used in subsequent statements  no-useless-assignment
  166:13  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\max-len.js
  344:25  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\max-lines.js
  112:13  error  This assigned value is not used in subsequent statements  no-useless-assignment
  123:13  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-dupe-class-members.js
  85:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-dupe-else-if.js
  103:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-empty-function.js
  43:9  error  This assigned value is not used in subsequent statements  no-useless-assignment
  76:9  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-extra-parens.js
  497:17  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-loss-of-precision.js
  67:17  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-object-constructor.js
  64:9  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\no-trailing-spaces.js
  130:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\utils\ast-utils.js
  1804:13  error  This assigned value is not used in subsequent statements  no-useless-assignment
  1805:13  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\rules\yield-star-spacing.js
  79:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\lib\source-code\token-store\index.js
  40:9  error  This assigned value is not used in subsequent statements  no-useless-assignment
  41:9  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\tests\lib\cli-engine\cli-engine.js
  2510:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2565:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2622:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\tests\lib\eslint\eslint.js
  2464:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2523:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2583:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\tests\lib\eslint\flat-eslint.js
  2400:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2465:21  error  This assigned value is not used in subsequent statements  no-useless-assignment
  2525:21  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\tests\lib\rule-tester\no-test-runners.js
  13:1  error  This assigned value is not used in subsequent statements  no-useless-assignment
  14:1  error  This assigned value is not used in subsequent statements  no-useless-assignment
  30:5  error  This assigned value is not used in subsequent statements  no-useless-assignment
  31:5  error  This assigned value is not used in subsequent statements  no-useless-assignment

C:\projects\tmp\ota-meshi-eslint\tools\update-readme.js
  152:9  error  This assigned value is not used in subsequent statements  no-useless-assignment
```

</details>

Most are initial assignments in declarations to default values like `null`, `0` or `""`. Perhaps these should be allowed in declarations (maybe optionally)?

There are also some that look to me like false positives for this rule, where the value is used on the right side of a subsequent assignment. For example:

https://github.com/eslint/eslint/blob/994596b07f5ff20a615a4be1ea03e5fd59cdb84b/Makefile.js#L375-L377

mdjermanovic commented:
This is another kind of false positive, I believe:

https://github.com/eslint/eslint/blob/994596b07f5ff20a615a4be1ea03e5fd59cdb84b/lib/linter/config-comment-parser.js#L102-L128

The error is reported on line 102, but if `JSON.parse()` throws, that `{}` will be used as return value on line 127.

ota-meshi commented:
Thank you for the review!

I think I fixed the false positives except for `try catch` and the initial value.
I'll work on the rest when I have time next time.

> Most are initial assignments in declarations to default values like null, 0 or "". Perhaps these should be allowed in declarations (maybe optionally)?

Yeah, I think we need an option to ignore the initial value.
However, some users may want to ignore `Number.MIN_SAFE_INTEGER` or `NaN`. I'm still not sure how many values to ignore. Do you think we should add an option to ignore all initial values and enable it by default?

mdjermanovic commented:
> I think I fixed the false positives except for `try catch` and the initial assignment.
> I'll work on the rest when I have time next time.

Testing on the eslint codebase now looks good to me, aside from the question about initial assignments. The `try-catch` case isn't a false positive, my mistake.

> > Most are initial assignments in declarations to default values like null, 0 or "". Perhaps these should be allowed in declarations (maybe optionally)?
> 
> Yeah, I think we need an option to ignore the initial assignments. However, some users may want to ignore `Number.MIN_SAFE_INTEGER`, `NaN`, `undefined`, `void 0`, or `Promise.resolve()`. I'm still not sure how many values to ignore. Do you think we should add an option to ignore all initial assignments and enable it by default?

I personally never use initial assignments when that value won't be used, but by just looking at our codebase it seems to be a relatively common practice. I'm not sure what would be the best solution. Even if the user wants to allow initial assignments like `let a = 0`, catching unused `let a = calculateSomethingExpensive()` seems useful.

Maybe there should be a boolean option for whether to check initial assignments (`false` by default?). When `true`, there could be another option to define an array of allowed values. Just not sure how it would work for things like `Promise.resolve()` (compare with the source code text literally?).

@nzakas what do you think about this?

nzakas commented:
I looked through some of the issues in the ESLint codebase found by this rule, and they look like good catches to me.

My feeling is that, at this point, I would ship this rule as-is without an additional option for initial assignments. Let's see what type of feedback we get once released and we can go from there. It will be a lot easier to make a decision on whether or not to add an option, and about what type of option values it should accept, when we get some real-world feedback.

In the meantime, I think we should enable this rule in `eslint-config-eslint` so we can dogfood it.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -0,0 +1,116 @@
+
+
+
+If the variable is not used after assignment is needless and most likely indicates that there is some behavior that the author wanted but didn't correctly do.
```
We definitely need more of a description here. We need to explain the concept of a "dead store" and why it's a problem. We also need to show an example and explain it.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -0,0 +1,116 @@
+rule_type: suggestion
+related_rules:
+- no-unused-vars
+---
```
Let's make sure to include more references for people. See the links in https://github.com/eslint/eslint/issues/17559#issuecomment-1732598182


We can use the "further_reading" key to list these out.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -0,0 +1,116 @@
+
+## When Not To Use It
+
+If you don't want to be notified about unnecessary arguments, you can safely turn this rule off.
```
Looks like a copy-paste error here.

```suggestion
If you don't want to be notified about values that are never read, you can safely disable this rule.
```

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,468 @@
+             * @param {Identifier} identifier The identifier to check.
+             * @returns {boolean} `true` if the identifier is used in the segment.
+             */
+            function isInSegment(segment, identifier) {
```
Can we do a bit of refactoring here to make this easier to understand? I'd find `isIdentifierUsedInSegment(identifier, segment)` clearer.

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,468 @@
+                );
+            }
+
+            /**
+             * Verify the given assignment info.
+             * @param {AssignmentInfo} targetAssignment The assignment info to verify.
+             * @param {AssignmentInfo[]} allAssignments The list of all assignment info for variables.
+             * @returns {void}
+             */
+            function verifyAssignment(targetAssignment, allAssignments) {
+
```
It's not clear to me what this function is doing. What does it mean to "verify assignment info"? Can you explain that further in the JSDoc, and maybe update the function name to better reflect what it's doing?

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,468 @@
+
+
+                /**
+                 * Gets the subsequent segments from the current segment.
```
I can't quite tell what "current segment" refers to here. Can you explain?

mdjermanovic commented:
> My feeling is that, at this point, I would ship this rule as-is without an additional option for initial assignments. Let's see what type of feedback we get once released and we can go from there. It will be a lot easier to make a decision on whether or not to add an option, and about what type of option values it should accept, when we get some real-world feedback.

I agree.

> In the meantime, I think we should enable this rule in `eslint-config-eslint` so we can dogfood it.

Do you mean to enable it in this PR? Sounds good to me.

nzakas commented:
> Do you mean to enable it in this PR? Sounds good to me.

Yes. This used to be standard practice whenever we added a new rule, but I think we got away from it at some point.

ota-meshi commented:
I've addressed some of the comments, but I found a false positive in the try statement, so I'm still working on it.

```js
let message = 'init'; // <- false positive
try {
    message = call().message;
} catch (e) {
    // ignore
}
console.log(message)
```

But the rule works fine if separate it into multiple statements like this 🤔 :

```js
let message = 'init';
try {
    const result = call();
    message = result.message;
} catch (e) {
    // ignore
}
console.log(message)
```

ota-meshi commented:
Thank you for your review!
I think I fixed the try-catch bug and the fix for your comments. Could you please check again?

nzakas commented on tests/conf/eslint-all.js:
```diff
@@ -31,7 +31,7 @@ describe("eslint-all", () => {
 
         assert.include(ruleNames, someRule);
-        assert.isBelow(count, 200);
+        assert.isAtLeast(count, 200);
```
Yeah, this is a result of us deprecating a bunch of rules. That put us under 200, but adding this new one puts us over again.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -3,11 +3,25 @@ title: no-useless-assignment
 - no-unused-vars
+further_reading:
+- https://en.wikipedia.org/wiki/Dead_store
+- https://www.npmjs.com/package/eslint-plugin-no-useless-assign
```
I think we can leave this off. We generally don't point to other plugins that do the same thing.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -3,11 +3,25 @@ title: no-useless-assignment
+- https://www.npmjs.com/package/eslint-plugin-no-useless-assign
+- https://sonarsource.atlassian.net/browse/RSPEC-1854
+- https://rules.sonarsource.com/javascript/RSPEC-1854/
+- https://rules.sonarsource.com/typescript/RSPEC-1854/
```
The previous three lines all basically point to the same information. I'd suggest we keep just the second one, which is JavaScript-specific.

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -3,11 +3,25 @@ title: no-useless-assignment
+
+"Dead stores" waste processing and memory, so it is better to remove unnecessary assignments to variables.
+
+Also, if the author intended the variable to be used, there is likely a mistake around the dead store.
```
We still need more in this section. Please show an example.

nzakas reviewed (approved):
Just a few suggestions for the docs but otherwise LGTM.

fasttime commented on docs/src/rules/no-useless-assignment.md:
````diff
@@ -0,0 +1,145 @@
+    }
+    console.log(v);
+}
+```
````
As it is now, the code in this example cannot be parsed in the Playground with the default `{ "sourceType": "module" }` because of the duplicated function name `foo` ([see here](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVzZWxlc3MtYXNzaWdubWVudDogXCJlcnJvclwiICovXG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBkb1NvbWV0aGluZyh2KTtcbiAgICB2ID0gJ3VudXNlZCc7XG59XG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIHYgPSAndW51c2VkJztcbiAgICAgICAgcmV0dXJuXG4gICAgfVxuICAgIGRvU29tZXRoaW5nKHYpO1xufVxuXG5mdW5jdGlvbiBmb28oKSB7XG4gICAgbGV0IHYgPSAndXNlZCc7XG4gICAgaWYgKGNvbmRpdGlvbikge1xuICAgICAgICBkb1NvbWV0aGluZyh2KTtcbiAgICB9IGVsc2Uge1xuICAgICAgICB2ID0gJ3VudXNlZCc7XG4gICAgfVxufVxuXG5mdW5jdGlvbiBmb28oKSB7XG4gICAgbGV0IHYgPSAndW51c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIHYgPSAndXNlZCc7XG4gICAgICAgIGRvU29tZXRoaW5nKHYpO1xuICAgICAgICByZXR1cm5cbiAgICB9XG59XG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgICBsZXQgdiA9ICd1c2VkJztcbiAgICBpZiAoY29uZGl0aW9uKSB7XG4gICAgICAgIGxldCB2ID0gJ3VzZWQnO1xuICAgICAgICBjb25zb2xlLmxvZyh2KTtcbiAgICAgICAgdiA9ICd1bnVzZWQnO1xuICAgIH1cbiAgICBjb25zb2xlLmxvZyh2KTtcbn0iLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)). This can be easily fixed by giving the functions unique names. That's what we did in https://github.com/eslint/eslint/pull/17633/files in other places. Since this is a new rule, it would be good to have the examples working in the Playground right from the start.

fasttime commented on docs/src/rules/no-useless-assignment.md:
````diff
@@ -0,0 +1,145 @@
+        v = 'used in next iteration';
+    }
+}
+```
````
Same as above, the function names in this code block should be unique.

ota-meshi commented:
Thank you for checking this PR. I have fixed the docs.

mdjermanovic commented on docs/src/rules/no-useless-assignment.md:
````diff
@@ -0,0 +1,146 @@
+    }
+}
+```
+
````
```suggestion

:::

```

To close `::: correct`.

ota-meshi replied:
Thank you!

mdjermanovic commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,473 @@
+
+                    // Checks if it is used in the same segment as the target assignment.
+                    if (
+                        targetAssignment.identifier.range[0] < reference.identifier.range[0] &&
```
I think this is a false negative due to `x` on the right side of `=` being considered as an usage of that assignment:

```js
/* eslint no-useless-assignment: "error" */

var x = 1;

x = x + 1; // false negative?

x = 5;

f(x);

```

nzakas commented on docs/src/rules/no-useless-assignment.md:
```diff
@@ -22,6 +19,10 @@ further_reading:
 "Dead stores" waste processing and memory, so it is better to remove unnecessary assignments to variables.
 
 Also, if the author intended the variable to be used, there is likely a mistake around the dead store.
+For example,
```
Sorry, I don't think I was clear. I'm looking for a code example in this section, something like:

```js
let id = "x1234";    // this is a "dead store" - this value is never read

id = generateId();

doSomethingWith(id);
```

That way, we can explain that `"x1234"` is a value that can't possibly be used, and then we say this type of error can occur while refactoring.

ota-meshi commented:
Thank you for checking this PR!
I changed this PR. So could you please check again?

mdjermanovic commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,496 @@
+                }
+                segmentInfo.last = node;
+            },
+            ":matches(VariableDeclarator[init!=null], AssignmentExpression, UpdateExpression):exit"(node) {
```
Some false positives:

```js
/* eslint no-useless-assignment: "error" */

const obj = { a: 5 };
const { a, b = a } = obj; // false positive on `a`
console.log(b); // 5

const arr = [6];
const [c, d = c] = arr; // false positive on `c`
console.log(d); // 6

```

I believe the cause is that read references appear in segments that are previous to segments that are current at the point when VariableDeclarator nodes are exited.

For example, `const { x = y } = obj`:

![image](https://github.com/eslint/eslint/assets/44349756/157b7636-9d97-4b5d-8766-664c4a8184c6)

ota-meshi replied:
I think I fixed that false positives. Could you please check the changes?

nzakas commented on tests/lib/rules/no-useless-assignment.js:
Can you switch this to use `FlatRuleTester` instead? We are going through and updating all rule tests to do so. It should look like this:

```js
const RuleTester = require("../../../lib/rule-tester/flat-rule-tester");
```

ota-meshi replied:
I changed the test case to use FlatRuleTester 👍

nzakas commented:
@mdjermanovic this is ready for you to check on your suggestions.

mdjermanovic commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,564 @@
+                     * If additional iterations are needed, consume this information and iterate.
+                     * @type {CodePathSegment[]}
+                     */
+                    queueSegments: [...targetAssignment.segment.nextSegments]
```
`nextSegments` of a `segment` never include unreachable segments, which is great as they should be skipped when we're analyzing a reachable `segment`.

However, when the `segment` itself is unreachable, analysis based on `nextSegments` would be wrong. For example:

```js
/* eslint no-useless-assignment: "error" */

() => {
    return;

    const x = 1; // The rule reports here, but this is unreachable. If this were reachable, the assignment would be valid.
    if (foo) {
        bar(x);
    }
}

```

I think the rule should just ignore assignments in unreachable segments.

ota-meshi commented on lib/rules/no-useless-assignment.js:
```diff
@@ -0,0 +1,564 @@
+                    }
+
+                    // Variables exported by "exported" block comments
+                    if (variable.eslintExported) {
```
I should use `variable.eslintUsed` as you say 👍

ota-meshi commented:
Thank you for your suggestions and also for finding the false positives. I fixed them.

nzakas commented on lib/rules/no-useless-assignment.js:
```diff
@@ -325,7 +325,10 @@ module.exports = {
-                        if (assignment === targetAssignment || assignment.segment !== targetAssignment.segment) {
+                        if (
+                            assignment === targetAssignment ||
+                            assignment.segments.every(segment => !targetAssignment.segments.includes(segment))
```
`every()` returns true when an array is empty. 
```suggestion
                            assignment.segments.length && assignment.segments.every(segment => !targetAssignment.segments.includes(segment))
```

nzakas reviewed (approved):
LGTM. Leaving open for @mdjermanovic to review.

mdjermanovic commented:
LGTM, I'll just close-reopen to check CI before merging.

mdjermanovic reviewed (approved):
Great work, thanks!

Given the complexity of the problem and JS syntax, I wouldn't be surprised if some edge cases pop up in production, but can't see any at the moment.

Sorry for the delay in reviewing, there have been many tasks for the v9 release lately.
